src/libfetchers/path.cc

```diff
@@ -82,18 +82,34 @@ struct PathInputScheme : InputScheme
 
     std::pair<Tree, Input> fetch(ref<Store> store, const Input & input) override
     {
+        std::string absPath;
         auto path = getStrAttr(input.attrs, "path");
 
-        // FIXME: check whether access to 'path' is allowed.
+        if (path[0] != '/' && input.parent)
+        {
+            auto parent = canonPath(*input.parent);
+
+            // the path isn't relative, prefix it
+            absPath = canonPath(parent + "/" + path);
 
-        auto storePath = store->maybeParseStorePath(path);
+            // for security, ensure that if the parent is a store path, it's inside it
+            if (!parent.rfind(store->storeDir, 0) && absPath.rfind(store->storeDir, 0))
+                throw BadStorePath("relative path '%s' points outside of its parent's store path %s, this is a security violation", path, parent);
+        }
```

Comment:
Hm, it looks like this causes subflakes that reference the parent flake to error out.

For example, https://github.com/divnix/digga/blob/56f94dbdaaac5d3c63eed1def0fc4760ec447a38/examples/groupByConfig/flake.nix#L8  implements a test flake for the "parent" flake library.

I guess we now would have to decide if this particular case is a valid use case.

I guess this error was implemented to fairly advertise the `import ../../`, which indeed is illegal. (Nice that you fixed this!)

Reply:
Here is a complete error output on our CI for additional context:
```console
$ check-downstream --show-trace
warning: Git tree '/home/blaggacao/ghq/github.com/divnix/digga/examples/downstream' is dirty
error: relative path '../../' points outside of its parent's store path /nix/store/24vs6gdb680393b0aidh4fyp3ijyshd7-source, this is a security violation

       … while fetching the input 'path:../../'

       … while updating the flake input 'digga'

       … while updating the lock file of flake 'git+file:///home/blaggacao/ghq/github.com/divnix/digga/examples/downstream'
exit 1: command [/nix/store/j3mddzlpyapqhx1443dp5hq707w92dkj-nix-2.4pre20210601_5985b8b/bin/nix flake lock --update-input digga "$@"] failed
```

Reply:
This error seems more contrived than I initially thought.

It seems that `input.parent` condition is triggered only when I replace the `digga` reference of the `downstream` submodule https://github.com/divnix/digga/tree/56f94dbdaaac5d3c63eed1def0fc4760ec447a38/examples with `path:../../` but the error doesn't occur, with https://github.com/divnix/digga/blob/56f94dbdaaac5d3c63eed1def0fc4760ec447a38/examples/groupByConfig/flake.nix#L8

I have to find out more about `input.parent`, to give a better error report.

Reply:
I analyzed a bit further and the impression I'm getting is that happens only when a git submodule boundary is crossed. //cc @nrdxp

I guess, this is another pair of shoes and not part of this PR.